Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.kt

```diff
@@ -0,0 +1,603 @@
+                    // The activity is being destroyed, so we can't use it to display an AlertMessage.
+                    // Fall back to a toast.
+                    Toast.makeText(emulationActivity, text, Toast.LENGTH_LONG).show()
+                    NotifyAlertMessageLock()
+                } else {
+                    AlertMessage.newInstance(caption, text, yesNo, isWarning)
+                        .show(fragmentManager, "AlertMessage")
+                }
+            }
+
+            // Wait for the lock to notify that it is complete.
+            try {
+                alertMessageSemaphore.acquire()
+            } catch (ignored: InterruptedException) {
+                Thread.currentThread().interrupt()
+            }
+
+            if (yesNo) {
+                result = AlertMessage.getAlertResult()
+            }
+        }
+
+        isShowingAlertMessage = false
+        return result
+    }
+
+    @JvmStatic
+    fun IsShowingAlertMessage(): Boolean = isShowingAlertMessage
+
+    @JvmStatic
+    fun NotifyAlertMessageLock() {
+        alertMessageSemaphore.release()
+    }
+
+    @JvmStatic
+    fun setEmulationActivity(emulationActivity: EmulationActivity) {
+        Log.verbose("[NativeLibrary] Registering EmulationActivity.")
+        emulationActivityRef = WeakReference(emulationActivity)
+    }
+
+    @JvmStatic
+    fun clearEmulationActivity() {
+        Log.verbose("[NativeLibrary] Unregistering EmulationActivity.")
+        emulationActivityRef.clear()
+    }
+
+    @JvmStatic
+    @Keep
+    fun finishEmulationActivity() {
+        val emulationActivity = emulationActivityRef.get()
+        if (emulationActivity == null) {
+            Log.warning("[NativeLibrary] EmulationActivity is null.")
+        } else {
+            Log.verbose("[NativeLibrary] Finishing EmulationActivity.")
+            emulationActivity.runOnUiThread(emulationActivity::finish)
+        }
+    }
+
+    @JvmStatic
+    @Keep
+    fun updateTouchPointer() {
+        val emulationActivity = emulationActivityRef.get()
+        if (emulationActivity == null) {
+            Log.warning("[NativeLibrary] EmulationActivity is null.")
+        } else {
+            emulationActivity.runOnUiThread(emulationActivity::initInputPointer)
+        }
+    }
+
+    @JvmStatic
+    @Keep
+    fun onTitleChanged() {
+        val emulationActivity = emulationActivityRef.get()
+        if (emulationActivity == null) {
+            Log.warning("[NativeLibrary] EmulationActivity is null.")
+        } else {
+            emulationActivity.runOnUiThread(emulationActivity::onTitleChanged)
+        }
+    }
+
+    @JvmStatic
+    @Keep
+    fun getRenderSurfaceScale(): Float {
+        val emulationActivity = emulationActivityRef.get()
+        if (emulationActivity == null) {
+            Log.warning("[NativeLibrary] EmulationActivity is null.")
+            return Resources.getSystem().displayMetrics.scaledDensity
+        }
```

Comment:
This change is also a bit bigger than I prefer to see in a "translate to Kotlin" commit. This one is uncontroversial enough that it's not a big problem, though.